File: ViewModels/TracksViewModel.swift
import SwiftUI
import Combine

/// ViewModel for the Tracks mixer screen
/// Acts as a lightweight wrapper around LooperViewModel for track-specific operations
@MainActor
final class TracksViewModel: ObservableObject {
    
    // MARK: - Dependencies
    
    private let looperVM: LooperViewModel
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Published State
    
    @Published private(set) var tracks: [Track] = []
    @Published private(set) var isPlaying: Bool = false
    @Published private(set) var isPaused: Bool = false
    @Published private(set) var currentPosition: Double = 0
    @Published var selectedTrackId: UUID? = nil
    @Published var showingQuantizeSheet = false
    @Published var trackToQuantize: Track? = nil
    @Published var showingInstrumentPicker = false
    @Published var trackToChangeInstrument: Track? = nil
    @Published var selectedTrackForFocus: Track? = nil  // For presenting MIDI editor
    
    // MARK: - Computed Properties
    
    var anyTrackSoloed: Bool {
        looperVM.anyTrackSoloed
    }
    
    var loopLengthBeats: Double {
        looperVM.loopLengthBeats
    }
    
    var bpm: Double {
        looperVM.bpm
    }
    
    // MARK: - Initialization
    
    init(looperVM: LooperViewModel) {
        self.looperVM = looperVM
        
        // Forward tracks from looperVM
        looperVM.$tracks
            .receive(on: DispatchQueue.main)
            .assign(to: &$tracks)
        
        // Forward isPlaying
        looperVM.$isPlaying
            .receive(on: DispatchQueue.main)
            .sink { [weak self] playing in
                self?.isPlaying = playing
            }
            .store(in: &cancellables)
        
        // Forward isPaused
        looperVM.$isPaused
            .receive(on: DispatchQueue.main)
            .sink { [weak self] paused in
                self?.isPaused = paused
            }
            .store(in: &cancellables)
        
        // Forward currentPosition for progress bar
        looperVM.$currentPosition
            .receive(on: DispatchQueue.main)
            .sink { [weak self] position in
                self?.currentPosition = position
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Track Operations
    
    func toggleMute(_ track: Track) {
        looperVM.toggleMute(track)
    }
    
    func toggleSolo(_ track: Track) {
        looperVM.toggleSolo(track)
    }
    
    func toggleLoop(_ track: Track) {
        looperVM.toggleLoop(track)
    }
    
    func setVolume(_ track: Track, volume: Float) {
        looperVM.setTrackVolume(track, volume: volume)
    }
    
    func setInstrument(_ track: Track, instrument: Instrument) {
        looperVM.setTrackInstrument(track, instrument: instrument)
    }
    
    func deleteTrack(_ track: Track) {
        looperVM.deleteTrack(track)
    }
    
    func quantizeTrack(_ track: Track, division: QuantizeDivision) {
        looperVM.quantizeTrack(track, division: division)
    }
    
    // MARK: - Quantize Flow
    
    func requestQuantize(_ track: Track) {
        trackToQuantize = track
        showingQuantizeSheet = true
    }
    
    func applyQuantize(division: QuantizeDivision) {
        guard let track = trackToQuantize else { return }
        quantizeTrack(track, division: division)
        showingQuantizeSheet = false
        trackToQuantize = nil
    }
    
    func cancelQuantize() {
        showingQuantizeSheet = false
        trackToQuantize = nil
    }
    
    // MARK: - Instrument Change Flow
    
    func requestInstrumentChange(_ track: Track) {
        trackToChangeInstrument = track
        showingInstrumentPicker = true
    }
    
    func applyInstrumentChange(instrument: Instrument) {
        guard let track = trackToChangeInstrument else { return }
        setInstrument(track, instrument: instrument)
        showingInstrumentPicker = false
        trackToChangeInstrument = nil
    }
    
    func cancelInstrumentChange() {
        showingInstrumentPicker = false
        trackToChangeInstrument = nil
    }
    
    // MARK: - Track Focus Navigation
    
    func selectTrackForFocus(_ track: Track) {
        selectedTrackForFocus = track
    }
    
    func closeTrackFocus() {
        selectedTrackForFocus = nil
    }
    
    var looperViewModel: LooperViewModel {
        looperVM
    }
    
    func getTrack(withId id: UUID) -> Track? {
        tracks.first { $0.id == id }
    }
}

File: android-app/feature/tracks/src/main/java/com/loopa/feature/tracks/TracksViewModel.kt
package com.loopa.feature.tracks

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.loopa.core.model.*
import com.loopa.feature.looper.LooperViewModel

/**
 * Lightweight mixer-screen adapter around LooperViewModel.
 * Ported from iOS TracksViewModel.swift.
 */
class TracksViewModel(val looperViewModel: LooperViewModel) {

    var selectedTrackId by mutableStateOf<String?>(null)
    var showingQuantizeSheet by mutableStateOf(false)
    var trackToQuantize by mutableStateOf<Track?>(null)
    var showingInstrumentPicker by mutableStateOf(false)
    var trackToChangeInstrument by mutableStateOf<Track?>(null)
    var selectedTrackForFocus by mutableStateOf<Track?>(null)

    val tracks: List<Track> get() = looperViewModel.tracks
    val isPlaying: Boolean get() = looperViewModel.isPlaying
    val isPaused: Boolean get() = looperViewModel.isPaused
    val currentPosition: Double get() = looperViewModel.currentPosition
    val anyTrackSoloed: Boolean get() = looperViewModel.anyTrackSoloed
    val loopLengthBeats: Double get() = looperViewModel.loopLengthBeats
    val bpm: Double get() = looperViewModel.bpm

    fun toggleMute(track: Track) = looperViewModel.toggleMute(track)
    fun toggleSolo(track: Track) = looperViewModel.toggleSolo(track)
    fun toggleLoop(track: Track) = looperViewModel.toggleLoop(track)
    fun setVolume(track: Track, volume: Float) = looperViewModel.setTrackVolume(track, volume)
    fun setInstrument(track: Track, instrument: Instrument) = looperViewModel.setTrackInstrument(track, instrument)
    fun deleteTrack(track: Track) = looperViewModel.deleteTrack(track)
    fun quantizeTrack(track: Track, division: QuantizeDivision) = looperViewModel.quantizeTrack(track, division)

    fun requestQuantize(track: Track) {
        trackToQuantize = track
        showingQuantizeSheet = true
    }

    fun applyQuantize(division: QuantizeDivision) {
        trackToQuantize?.let { quantizeTrack(it, division) }
        showingQuantizeSheet = false
        trackToQuantize = null
    }

    fun cancelQuantize() {
        showingQuantizeSheet = false
        trackToQuantize = null
    }

    fun requestInstrumentChange(track: Track) {
        trackToChangeInstrument = track
        showingInstrumentPicker = true
    }

    fun applyInstrumentChange(instrument: Instrument) {
        trackToChangeInstrument?.let { setInstrument(it, instrument) }
        showingInstrumentPicker = false
        trackToChangeInstrument = null
    }

    fun cancelInstrumentChange() {
        showingInstrumentPicker = false
        trackToChangeInstrument = null
    }

    fun selectTrackForFocus(track: Track) {
        selectedTrackForFocus = track
    }

    fun closeTrackFocus() {
        selectedTrackForFocus = null
    }

    fun getTrack(withId: String): Track? = tracks.firstOrNull { it.id == withId }
}
